redis/client.go

```diff
@@ -0,0 +1,219 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/coinbase/memcachedbetween/pool"
+	"github.com/coinbase/mongobetween/util"
+	"github.com/coinbase/redisbetween/utils"
+	"go.uber.org/zap"
+	"io"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Options struct {
+	Addr         string
+	Database     int
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+	Readonly     bool
+	MinPoolSize  int
+	MaxPoolSize  int
+}
+
+type Client struct {
+	server    pool.ServerWrapper
+	messenger Messenger
+	config    Options
+}
+
+type ClientInterface interface {
+	Address() string
+	Call(ctx context.Context, msg []*Message) ([]*Message, error)
+	CheckoutConnection(ctx context.Context) (conn pool.ConnectionWrapper, err error)
+	Close(ctx context.Context) error
+}
+
+// NewClient opens a connection pool to the redis server
+// and returns a handle.
+func NewClient(ctx context.Context, c *Options) (ClientInterface, error) {
+	config := *c
+	server, err := connectToServer(ctx, &config)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		server:    server,
+		messenger: WireMessenger{},
+		config:    config,
+	}, nil
+}
+
+func (r *Client) Address() string {
+	return r.config.Addr
+}
+
+// Call performs a round trip call to the redis upstream and
+// returns the response message sent by the server. The method
+// will fail if the command is blocking
+func (r *Client) Call(ctx context.Context, msg []*Message) ([]*Message, error) {
+	log := ctx.Value(utils.CtxLogKey).(*zap.Logger)
+
+	var err error
+	var conn pool.ConnectionWrapper
+	if conn, err = r.CheckoutConnection(ctx); err != nil {
+		return nil, err
+	}
+
+	if s, ok := ctx.Value(utils.CtxStatsdKey).(statsd.ClientInterface); ok {
+		id, address := conn.ID(), conn.Address().String()
+		defer func(start time.Time) {
+			_ = s.Timing("upstream_round_trip", time.Since(start), []string{
+				fmt.Sprintf("upstream_id:%d", id),
+				fmt.Sprintf("address:%s", address),
+				fmt.Sprintf("success:%v", err == nil),
+			}, 1)
+		}(time.Now())
+	}
+
+	l := log.With(zap.Uint64("upstream_id", conn.ID()))
```

Comment:
I get why you want to add the `upstream` tag before calling the defer function, but I think `defer` should be moved back to right after we check out the connection. Currently there's nothing that could go wrong in lines 75-86 that could cause us to panic and return before checking the connection back in, but it's safer to keep the checkout/defer/return right beside each other in case there's a change to this code in the future that could cause us to return before checking the connection back in.

Reply:
done